File: io/lucamienert/basic/ast/BranchStatement.java
package io.lucamienert.basic.ast;

import io.lucamienert.basic.instruction.Instruction;
import io.lucamienert.basic.instruction.InstructionSequence;
import io.lucamienert.basic.instruction.OpCode;

public final class BranchStatement extends Statement {

    private final BranchType type;
    private final int target;

    public BranchStatement(BranchType type, int target) {
        this.type = type;
        this.target = target;
    }

    public BranchType getType() {
        return type;
    }

    public int getTarget() {
        return target;
    }

    @Override
    public void compile(InstructionSequence seq) {
        OpCode opcode;
        switch (type) {
            case GOTO:
                opcode = OpCode.JMP;
                break;
            case GOSUB:
                opcode = OpCode.CALL;
                break;
            default:
                throw new AssertionError();
        }
        seq.append(new Instruction(opcode, seq.createLineLabel(target)));
    }
}

File: BasicCompiler/Ast/BranchStatement.cs
using BasicCompiler.Instructions;

namespace BasicCompiler.Ast;

public class BranchStatement : Statement
{
    public BranchType BranchType { get; set; }
    public int Target { get; set; }

    public BranchStatement(BranchType type, int target)
    {
        BranchType = type;
        Target = target;
    }

    public override void Compile(InstructionSequence instructionSequence)
    {
        var opcode = BranchType switch
        {
            BranchType.GOTO => OpCode.JMP,
            BranchType.GOSUB => OpCode.CALL,
            _ => throw new AssertionException(),
        };

        instructionSequence.Append(new Instruction(opcode, InstructionSequence.CreateLineLabel(Target)));
    }
}
